UnchainedLauncherCore/src/API/RegisteredServer.cs

```diff
@@ -0,0 +1,190 @@
+﻿using Newtonsoft.Json.Linq;
+using Octokit;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Threading;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using UnchainedLauncher.Core.API;
+using UnchainedLauncher.Core.JsonModels.Metadata.V3;
+using LanguageExt.Pipes;
+
+namespace UnchainedLauncher.Core.API
+{
+
+    public static class TaskExtensions{
+        public static Task WhenCanceled(this CancellationToken cancellationToken)
+        {
+            var tcs = new TaskCompletionSource<bool>();
+#pragma warning disable //There are no nulls that can emerge here
+            cancellationToken.Register(s => ((TaskCompletionSource<bool>)s).SetResult(true), tcs);
+#pragma warning restore
+            return tcs.Task;
+        }
+
+    }
+
+    //TODO? Implement IObservableProperty here for GUI bindings
+    public class RegisteredServer : IDisposable
```

Comment:
Our "FodyWeavers" library does some macros/codegen which makes this really easy. You just need to add this annotation to the class which you want to provide update notifications. The `INotifyPropertyChanged` interface is what provides notifications back to the gui.

https://github.com/Chiv2-Community/UnchainedLauncher/blob/main/C2GUILauncher/src/ViewModels/ServerLauncherViewModel.cs#L24